wp-includes/class-wp-service-workers.php

```diff
@@ -0,0 +1,216 @@
+<?php
+/**
+ * Dependencies API: WP_Service_Workers class
+ *
+ * @since ?
+ *
+ * @package PWA
+ */
+
+/**
+ * Class used to register service workers.
+ *
+ * @since ?
+ *
+ * @see WP_Dependencies
+ */
+class WP_Service_Workers extends WP_Scripts {
+
+	/**
+	 * Param for service workers.
+	 *
+	 * @var string
+	 */
+	public $query_var = 'wp_service_worker';
+
+	/**
+	 * Output for service worker scope script.
+	 *
+	 * @var string
+	 */
+	public $output = '';
+
+	/**
+	 * WP_Service_Workers constructor.
+	 */
+	public function __construct() {
+		parent::__construct();
+		global $wp_filesystem;
+
+		if ( ! class_exists( 'WP_Filesystem' ) ) {
+			require_once ABSPATH . '/wp-admin/includes/file.php';
+		}
+
+		if ( null === $wp_filesystem ) {
+			WP_Filesystem();
+		}
+	}
+
+	/**
+	 * Initialize the class.
+	 */
+	public function init() {
+		/**
+		 * Fires when the WP_Service_Workers instance is initialized.
+		 *
+		 * @param WP_Service_Workers $this WP_Service_Workers instance (passed by reference).
+		 */
+		do_action_ref_array( 'wp_default_service_workers', array( &$this ) );
+	}
+
+	/**
+	 * Register service worker.
+	 *
+	 * Registers service worker if no item of that name already exists.
+	 *
+	 * @param string $handle Name of the item. Should be unique.
+	 * @param string $path   Path of the item relative to the WordPress root directory.
+	 * @param array  $deps   Optional. An array of registered item handles this item depends on. Default empty array.
+	 * @param bool   $ver    Always false for service worker.
+	 * @param mixed  $scope  Optional. Scope of the service worker. Default relative path.
+	 * @return bool Whether the item has been registered. True on success, false on failure.
+	 */
+	public function add( $handle, $path, $deps = array(), $ver = false, $scope = null ) {
+
+		// Set default scope if missing.
+		if ( ! $scope ) {
+			$scope = site_url( '/', 'relative' );
+		}
+
+		if ( false === parent::add( $handle, $path, $deps, false, $scope ) ) {
```

Comment:
Instead of passing `$scope` as the args, I think it should pass `compact( 'scope' )`. This will allow other args to be added in the future in addition to the scope.

Also, maybe a script should be able to added to multiple scopes as I mentioned above, so maybe it should be `scopes` instead.